modules/fedimint-wallet-client/src/pegin_monitor.rs

```diff
@@ -0,0 +1,488 @@
+
+        for outcome in outcomes {
+            min = match (min, outcome.retry_delay()) {
+                (None, time) => time,
+                (Some(min), None) => Some(min),
+                (Some(min), Some(time)) => Some(cmp::min(min, time)),
+            };
+        }
+
+        min
+    }
+
+    fn get_claimed_now_outpoints(outcomes: &[CheckOutcome]) -> Vec<bitcoin::OutPoint> {
+        let mut res = vec![];
+        for outcome in outcomes {
+            if let CheckOutcome::Claimed { outpoint } = outcome {
+                res.push(*outpoint);
+            }
+        }
+
+        res
+    }
+}
+
+/// Query via btc rpc for a history of an address derived with `tweak_idx` and
+/// claim any peg-ins that are ready.
+///
+/// Return a list of [`CheckOutcome`]s for each matching output.
+#[instrument(skip_all, fields(tweak_idx))]
+async fn check_idx_pegins(
+    data: &WalletClientModuleData,
+    tweak_idx: TweakIdx,
+    btc_rpc: &DynBitcoindRpc,
+    module_rpc: &DynModuleApi,
+    db: &Database,
+    client_ctx: &ClientContext<WalletClientModule>,
+) -> Result<Vec<CheckOutcome>, anyhow::Error> {
+    let current_consensus_block_count = module_rpc.fetch_consensus_block_count().await?;
+    let (script, address, tweak_key, operation_id) = data.derive_peg_in_script(tweak_idx);
+    btc_rpc.watch_script_history(&script).await?;
+
+    let history = btc_rpc.get_script_history(&script).await?;
+
+    debug!(target: LOG_CLIENT_MODULE_WALLET, %address, num_txes=history.len(), "Got history of a peg-in address");
+
+    let mut outcomes = vec![];
+
+    for (transaction, out_idx) in filter_onchain_deposit_outputs(history.into_iter(), &script) {
+        let txid = transaction.txid();
+        let outpoint = bitcoin::OutPoint {
+            txid,
+            vout: out_idx,
+        };
+
+        let claimed_peg_in_key = ClaimedPegInKey {
+            peg_in_index: tweak_idx,
+            btc_out_point: outpoint,
+        };
+
+        if db
+            .begin_transaction_nc()
+            .await
+            .get_value(&claimed_peg_in_key)
+            .await
+            .is_some()
+        {
+            debug!(target: LOG_CLIENT_MODULE_WALLET, %txid, %out_idx, "Already claimed");
+            outcomes.push(CheckOutcome::AlreadyClaimed);
+            continue;
+        }
+        let finality_delay = u64::from(data.cfg.finality_delay);
+
+        let Some(tx_block_height) = btc_rpc.get_tx_block_height(&txid).await? else {
+            outcomes.push(CheckOutcome::Pending {
+                num_blocks_needed: finality_delay,
+            });
+            debug!(target: LOG_CLIENT_MODULE_WALLET, %txid, %out_idx, "In the mempool");
+            continue;
+        };
+
+        // if a tx was confirmed in block N and N+1 is a current block count, it has 1
+        // confirmation, and 0 extra confirmations
+        let num_blocks_needed = tx_block_height
+            .saturating_add(1)
+            .saturating_sub(current_consensus_block_count);
```

Comment:
[nit]

No need to change, but the difference between block height and block count might be easier to understand using an intermediate variable `tx_block_count`, e.g.

```rust
let tx_block_count = match btc_rpc.get_tx_block_height(&txid).await? {
    Some(tx_block_height) => tx_block_height.saturating_add(1),
    None => {
        outcomes.push(CheckOutcome::Pending {
            num_blocks_needed: finality_delay,
        });
        debug!(target: LOG_CLIENT_MODULE_WALLET, %txid, %out_idx, "In the mempool");
        continue;
    }
};
let num_blocks_needed = tx_block_count.saturating_sub(current_consensus_block_count);
```

Reply:
I like it.